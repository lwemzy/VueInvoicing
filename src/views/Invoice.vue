<template>
  <v-container fluid id="create">
    <v-row>
      <v-col cols="2">
        <!-- customer card -->
        <v-card class="mx-auto" max-width="250">
          <v-img class="white--text" height="250px" :src="getCustomer.avatar"></v-img>

          <v-card-text class="text-center">
            <span>{{getCustomer.name}}</span>
          </v-card-text>

          <v-card-actions>
            <v-list-item>
              <v-row align="center" justify="end">
                <!-- create invoice modal -->
                <v-dialog v-model="dialog" persistent max-width="450px">
                  <template v-slot:activator="{ on }">
                    <v-btn text color="primary" x-small v-on="on">Create Invoice</v-btn>
                  </template>
                  <!-- invoice card -->
                  <InvoiceCard
                    :editedInvoice="editedInvoice"
                    :formTitle="formTitle"
                    :invoiceRules="invoiceRules"
                    :valid="valid"
                    :menu="menu"
                    @close="close"
                    @save="save"
                  ></InvoiceCard>
                  <!-- invoice card -->
                </v-dialog>
                <!-- end create invoice modal -->
              </v-row>
            </v-list-item>
          </v-card-actions>
        </v-card>
      </v-col>
      <v-col cols="10">
        <v-container>
          <v-row>
            <v-col cols="3" v-for="invoice in paginatedData" :key="invoice.id">
              <v-card max-width="200" class="mx-auto pa-2">
                <v-card-title class="subtitle-1">Invoice No: {{invoice.invoice_no}}</v-card-title>
                <v-card-text>
                  Created At: {{invoice.createdAt | formatDate}}
                  <br />
                  Due Date: {{invoice.due_date | formatDate}}
                </v-card-text>
                <v-card-actions>
                  <v-row align="center" justify="center">
                    <!-- invoice items modal -->
                    <InvoiceItems :customerId="id" :invoiceId="invoice.id"></InvoiceItems>
                    <!-- invoice items modal -->
                    <v-btn class="ma-1" outlined small fab color="grey" @click="editItem(invoice)">
                      <v-icon>mdi-pencil</v-icon>
                    </v-btn>
                    <v-btn
                      class="ma-1"
                      outlined
                      small
                      fab
                      color="grey"
                      @click="deleteInvoice({customerId: id, id:invoice.id})"
                    >
                      <v-icon>mdi-delete</v-icon>
                    </v-btn>
                  </v-row>
                </v-card-actions>
              </v-card>
            </v-col>
          </v-row>
        </v-container>
      </v-col>
    </v-row>
    <v-row justify="end">
      <div class="text-center pa-2 ma-2">
        <v-pagination
          @next="nextPage"
          @previous="prevPage"
          v-model="page"
          :length="paginatedPageCount"
          circle
        ></v-pagination>
      </div>
    </v-row>
    <v-snackbar bottom right v-model="snackbar" :timeout="timeout">{{ text }}</v-snackbar>
  </v-container>
</template>

<script>
import format from "date-fns/format";

import { mapActions, mapGetters } from "vuex";
import InvoiceItems from "../components/InvoiceItems";
import InvoiceCard from "../components/InvoiceCard";

export default {
  name: "Invoice",
  props: ["id"],
  components: {
    InvoiceItems,
    InvoiceCard
  },
  data() {
    return {
      snackbar: false,
      text: "",
      timeout: 2000,
      valid: true,
      invoiceRules: [
        v => !!v || "Address is required",
        v => /^\d+$/.test(v) || "Numbe a number"
      ],
      page: 1,
      pageCount: 0,
      itemsPerPage: 20,
      dialog: false,
      dialog2: false,
      dialog3: false,
      menu: false,
      editedIndex: -1,
      editedInvoice: {
        due_date: new Date().toISOString().substr(0, 10),
        invoice_no: ""
      },
      defaultInvoice: {
        due_date: new Date().toISOString().substr(0, 10),
        invoice_no: ""
      }
    };
  },
  computed: {
    ...mapGetters(["getCustomer", "allInvoices"]),
    formTitle() {
      return this.editedIndex === -1 ? "New Invoice" : "Edit Invoice";
    },
    paginatedData() {
      const start = this.pageCount * this.itemsPerPage;
      const end = start + this.itemsPerPage;
      return this.allInvoices.slice(start, end);
    },
    paginatedPageCount() {
      let l = this.allInvoices.length;
      let s = this.itemsPerPage;
      return Math.ceil(l / s);
    }
  },
  methods: {
    ...mapActions([
      "customer",
      "getInvoices",
      "newInvoice",
      "deleteInvoice",
      "editInvoice"
    ]),
    editItem(item) {
      this.editedIndex = this.allInvoices.findIndex(val => val.id === item.id);
      this.editedInvoice = Object.assign({}, item);
      this.dialog = true;
    },
    close() {
      this.dialog = false;
      setTimeout(() => {
        // clear modal on close
        this.editedInvoice = Object.assign({}, this.defaultInvoice);
        this.editedIndex = -1;
      }, 300);
    },

    save() {
      if (this.editedIndex > -1) {
        // edited customer
        const editedInvoice = Object.assign(this.editedInvoice, {
          customerId: this.id
        });
        this.editInvoice(editedInvoice);
        this.snackbar = true;
        this.text = "Invoice Successfully Edited";
      } else {
        // new invoice
        const invoice = Object.assign(this.editedInvoice, {
          customerId: this.id
        });
        this.newInvoice(invoice);
        this.snackbar = true;
        this.text = "Invoice Successfully Saved";
      }
      this.close();
    },

    nextPage() {
      this.pageCount++;
    },
    prevPage() {
      this.pageCount--;
    }
  },
  filters: {
    formatDate(date) {
      return format(new Date(date), "yyyy-MM-dd");
    }
  },
  created() {
    this.customer(this.id);
    this.getInvoices(this.id);
  }
};
</script>