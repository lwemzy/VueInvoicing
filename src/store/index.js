import Vue from 'vue';
import Vuex from 'vuex';
import Customers from './modules/customers';
import Invoice from './modules/invoice';
import InvoiceItems from './modules/invoiceItems';
import Products from './modules/products';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    Customers,
    Invoice,
    InvoiceItems,
    Products
  }
});
