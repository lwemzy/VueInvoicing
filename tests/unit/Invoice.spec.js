import { createLocalVue, mount } from '@vue/test-utils';
import Invoice from '@/views/Invoice.vue';
import Vue from 'vue';
import Vuetify from 'vuetify';
Vue.use(Vuetify);
import store from '@/store';

const localVue = createLocalVue();

describe('Invoice.vue', () => {
  let vuetify;

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  it('has a button', () => {
    const wrapper = mount(Invoice, {
      store,
      localVue,
      vuetify
    });
    expect(wrapper.contains('button')).toBe(true);
  });

  it('Card should have text and match snapshot', () => {
    const wrapper = mount(Invoice, {
      store,
      localVue,
      vuetify
    });

    // create snapshot files of the HTML output
    expect(wrapper.html()).toMatchSnapshot();

    const text = wrapper.find('.v-card__text > span');

    expect(text.text()).toBe('');
  });

  it('has a Pagination', () => {
    const wrapper = mount(Invoice, {
      store,
      localVue,
      vuetify
    });
    expect(wrapper.html()).toMatchSnapshot();

    const pagination = wrapper.find('.v-pagination__navigation');

    expect(pagination.text()).toBe('');
  });

  it('button is clickable', () => {
    const wrapper = mount(Invoice, {
      store,
      localVue,
      vuetify
    });

    const button = wrapper.find('button');

    // Simulate a click on the button
    button.trigger('click');
  });
});
