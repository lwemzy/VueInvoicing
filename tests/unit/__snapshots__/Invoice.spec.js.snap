// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invoice.vue Card should have text and match snapshot 1`] = `
<div class="container container--fluid" id="create">
  <div class="row">
    <div class="col col-2">
      <div class="mx-auto v-card v-sheet theme--light" style="max-width: 250px;">
        <div class="v-responsive v-image white--text" style="height: 250px;">
          <div class="v-responsive__content"></div>
        </div>
        <div class="v-card__text text-center"><span></span></div>
        <div class="v-card__actions">
          <div tabindex="-1" class="v-list-item theme--light">
            <div class="row align-center justify-end">
              <div role="dialog" class="v-dialog__container"><button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--x-small primary--text"><span class="v-btn__content">Create Invoice</span></button>
                <div role="document" class="v-dialog__content" style="z-index: 0;">
                  <div class="v-dialog v-dialog--persistent" style="max-width: 450px; display: none;"></div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col col-10">
      <div class="container">
        <div class="row"></div>
      </div>
    </div>
  </div>
  <div class="row justify-end">
    <div class="text-center pa-2 ma-2">
      <ul class="v-pagination v-pagination--circle theme--light">
        <li><button type="button" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></button></li>
        <li><button type="button" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></button></li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Invoice.vue has a Pagination 1`] = `
<div class="container container--fluid" id="create">
  <div class="row">
    <div class="col col-2">
      <div class="mx-auto v-card v-sheet theme--light" style="max-width: 250px;">
        <div class="v-responsive v-image white--text" style="height: 250px;">
          <div class="v-responsive__content"></div>
        </div>
        <div class="v-card__text text-center"><span></span></div>
        <div class="v-card__actions">
          <div tabindex="-1" class="v-list-item theme--light">
            <div class="row align-center justify-end">
              <div role="dialog" class="v-dialog__container"><button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--x-small primary--text"><span class="v-btn__content">Create Invoice</span></button>
                <div role="document" class="v-dialog__content" style="z-index: 0;">
                  <div class="v-dialog v-dialog--persistent" style="max-width: 450px; display: none;"></div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col col-10">
      <div class="container">
        <div class="row"></div>
      </div>
    </div>
  </div>
  <div class="row justify-end">
    <div class="text-center pa-2 ma-2">
      <ul class="v-pagination v-pagination--circle theme--light">
        <li><button type="button" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></button></li>
        <li><button type="button" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></button></li>
      </ul>
    </div>
  </div>
</div>
`;
